lib/dataset/sized_dict.cpp

```diff
@@ -35,14 +41,52 @@ SizedDict<Key, Value>::SizedDict(Sizes sizes, holder_type items,
   m_readonly = readonly; // NOLINT(cppcoreguidelines-prefer-member-initializer)
 }
 
+namespace {
+void merge_sizes_into(Sizes &target, const Dimensions &s) {
+  using std::to_string;
+
+  for (const auto &dim : s) {
+    if (target.contains(dim)) {
+      const auto a = target[dim];
+      const auto b = s[dim];
+      if (a == b + 1) // had bin-edges, replace by regular coord
+        target.resize(dim, b);
+      else if (a + 1 == b) { // had regular coord, got extra by bin-edges
+        // keep current
+      } else if (a != b)
+        throw except::DimensionError(
+            "Conflicting length in dimension " + to_string(dim) + ": " +
+            to_string(target[dim]) + " vs " + to_string(s[dim]));
+    } else {
+      target.set(dim, s[dim]);
+    }
+  }
+}
+} // namespace
+
+template <class Key, class Value>
+SizedDict<Key, Value>::SizedDict(AutoSizeTag, holder_type items,
+                                 const bool readonly) {
+  Sizes sizes;
+  for (auto &&[key, value] : items) {
+    merge_sizes_into(sizes, value.dims());
+  }
+  m_sizes = std::move(sizes);
+
+  for (auto &&[key, value] : items)
+    set(key, std::move(value));
+  // `set` requires Dict to be writable, set readonly flag at the end.
+  m_readonly = readonly; // NOLINT(cppcoreguidelines-prefer-member-initializer)
+}
```

Comment:
```suggestion
                                 const bool readonly) : SizedDict(auto_sizes(items), items, readonly) {}
```
with an appropriate helper function `auto_sizes`, or am I missing some logic here?